### PR TITLE
[feature] Allow '~' pattern as alternative for '!' to exclude references

### LINKED ISSUE #13834: [bug] Bash doesn't like the use "!" when cleaning up old revisions
### Environment details

* Operating System+version: up-to-date Manjaro
* Conan version: 2.0.4
* Python version: 3.10.10

When trying to clean up old revisions using the advised command:
~~~~
> conan remove zlib/1.2.13#!latest
bash: !latest: event not found
> conan remove "zlib/1.2.13#!latest"
bash: !latest: event not found
> conan remove zlib/1.2.13\#\!latest
Remove the recipe and all the packages of 'zlib/1.2.13#13c96f538b52e1600c40b88994de240f'? (yes/no):
~~~~

So I think the choice of "#!" is unlucky on `bash`.

### Steps to reproduce

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by danimtb**
Changelog: Feature: Allow "~" symbol as a exclude pattern for references
Docs: missing

Close https://github.com/conan-io/conan/issues/13834

- [x] Refer to the issue that supports this Pull Request: closes https://github.com/conan-io/conan/issues/13834
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

### CODE REVIEW

**Review by AbrilRBS — APPROVED** [review]
Nice!

File: conan/cli/args.py [hunk 1]
```diff
@@ -15,6 +15,8 @@
                        pattern uses 'fnmatch' style wildcards.
     --build=![pattern] Excluded packages, which will not be built from the source, whose package
                        reference matches the pattern. The pattern uses 'fnmatch' style wildcards.
+    --build=~[pattern] Excluded packages, which will not be built from the source, whose package
```
Comment (memsharded) [inline conan/cli/args.py:16]: If we are going to converge to using ``~``, maybe remove the ``!`` from the docstring?

File: conans/client/graph/build_mode.py [hunk 2]
```diff
@@ -9,7 +9,7 @@ class BuildMode:
                    => ["hello/0.1@foo/bar"] if user wrote "--build hello/0.1@foo/bar"
                    => False if user wrote "never"
                    => True if user wrote "missing"
-                   => ["!foo"] means exclude when building all from sources
+                   => ["!foo"] or ["~foo"] means exclude when building all from sources
```
Comment (memsharded) [inline conans/client/graph/build_mode.py:12]: Same, remove legacy from docstrings

**Review by memsharded — APPROVED** [review]
Looks good!

File: conans/client/graph/build_mode.py (hunk 2 above)
Comment (danimtb) [reply conans/client/graph/build_mode.py:12]: If this is just code documentation (not displayed in the help commands), I'd prefer to keep it consistent with the code

**Comment by danimtb** [comment]
Do you think this notation should also be extended to `exports_sources` attribute? https://docs.conan.io/2/reference/conanfile/attributes.html#exports-sources

File: conans/client/graph/build_mode.py (hunk 2 above)
Comment (AbrilRBS) [reply conans/client/graph/build_mode.py:12]: I agree with Dani here 👍

Comment (memsharded) [reply conans/client/graph/build_mode.py:12]: Yes, sorry, I thought it was part of the things that get automatically published in docs from docstrings. Of course, as long as it doesn't appear in the user docs or in ``conan -h``, it is better to leave it.

**Review by memsharded — APPROVED** [review]
The most important part of this PR is not that it is green, but it has to be manually tested in all different terminals, Windows (Powershell, Cmd, Cmder, Git bash), Linux (Bash, hopefully some other common shell), OSX.

Could you please confirm it works? Thanks!

**Comment by danimtb** [comment]
### Windows
#### cmder
```
λ conan list zlib/1.2.8#!latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)

λ conan list zlib/1.2.8#~latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)

```

#### cmd
```
> conan list zlib/1.2.8#!latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)

> conan list zlib/1.2.8#~latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)
```

#### powershell
```
$ conan list zlib/1.2.8#!latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)

$ conan list zlib/1.2.8#~latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)
```

#### git bash
```
$ conan list zlib/1.2.8#!latest -r conancenter
C:\cmder\vendor\conemu-maximus5\..\git-for-windows\usr\bin\bash: !latest: event not found


λ conan list zlib/1.2.8#~latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)
```

### Linux (docker)

#### bash
```
$ conan list zlib/1.2.8#!latest -r conancenter
bash: !latest: event not found

$ conan list zlib/1.2.8#~latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)
```

#### zsh
```
% conan list zlib/1.2.8#!latest -r conancenter
zsh: event not found: latest

% conan list zlib/1.2.8#~latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)
```

### Macos

#### zsh
```
% conan conan list zlib/1.2.8#!latest -r conancenter
zsh: event not found: latest

% conan list zlib/1.2.8#~latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)
```

#### fish
```
$ conan list zlib/1.2.8#!latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)

$ conan list zlib/1.2.8#~latest -r conancenter
conancenter
  zlib
    zlib/1.2.8
      revisions
        675e943afacc57989da75fe34cccffe8 (2019-10-13 09:55:53 UTC)
        c6769fac6d479f594ffa32c970bf7106 (2019-12-19 09:02:52 UTC)
        83cfd504315595c68ffdc7b7eff853d7 (2020-03-13 17:13:37 UTC)
        b66e7f62a2b2532eb253176c0bef216e (2020-08-07 17:25:48 UTC)
        74ce22a7946b98eda72c5f8b5da3c937 (2020-09-18 11:03:11 UTC)
```

**Comment by memsharded** [comment]
Amazing job, lets merge it, thanks!!

**Pull request merged**